### ISSUE TITLE
Vectorized gcd and lcm?

### ISSUE THREAD

**Issue by MasonProtter** [JuliaSIMD/LoopVectorization.jl#64]
To support Matrices of `Rational` numbers in [Gaius.jl](https://github.com/MasonProtter/Gaius.jl) it would be really handy if we had vectorized greatest common denominator `gcd` and least common multiple `lcm` functions. Do you know of an easy way to support them in LoopVectorization?

**Comment by chriselrod** [comment]
SIMDPirates.jl would probably be the place to add them. [Here](https://github.com/JuliaLang/julia/blob/master/base/intfuncs.jl) are the Julia definitions.

I [just added](https://github.com/chriselrod/SIMDPirates.jl/commit/047e753b746310708692c7a8dcab069e75faa294) a vectorized trailing zeros function (will probably only be vectorized with AVX512CD, but I found my [rand-log-from-integer](https://github.com/chriselrod/VectorizedRNG.jl/blob/master/src/special_approximations.jl#L259) function using leading zeros was faster on AVX2 anyway).

Those while loops are going to hurt. It'll have to repeat until every element of the vector meets the condition.

You're welcome to try and see if you can come up with a good implementation.

**Comment by MasonProtter** [comment]
Thanks! Probably not a good candidate for vectorization then 😅

**Comment by chriselrod** [comment]
Probably not, but "works" tends to be better than "does not work".
On the master branches of SIMDPirates and VectorizationBase:
```julia
julia> function Base.gcd(a::SVec{W,T}, b::SVec{W,T}) where {W, T <: Integer}
           #a == 0 && return abs(b)
           #b == 0 && return abs(a)
           za = trailing_zeros(a)
           zb = trailing_zeros(b)
           k = min(za, zb)
           u = unsigned(abs(a >> za))
           v = unsigned(abs(b >> zb))
           while true
               incomplete = u != v;
               incomplete == SIMDPirates.VectorizationBase.mask(Val(W), 0) && break
               ugv = u > v
               unew = vifelse(ugv, v, u)
               v = vifelse(ugv, u, v); u = unew
               vnew = v - u
               v = vifelse(incomplete, vnew >> trailing_zeros(vnew), v)
           end
           r = u << k
           # T(r) would throw InexactError; we want OverflowError instead
           #r > typemax(T) && __throw_gcd_overflow(a, b)
           r % T
       end

julia> v1 = SVec(Core.VecElement.((6,12, 24, 51)))
SVec{4,Int64}<6, 12, 24, 51>

julia> v2 = SVec(Core.VecElement.((9, 8, 18, 34)))
SVec{4,Int64}<9, 8, 18, 34>

julia> gcd(v1, v2)
SVec{4,Int64}<3, 4, 6, 17>
```

**Comment by chriselrod** [comment]
Long time coming, but fixed with LoopVectorization 0.12.
```julia
julia> x = Int.(rand(Int16, 1024));

julia> y = Int.(rand(Int16, 1024));

julia> lcm.(x,y) == @avx abs.(lcm.(x,y))
true
```
Although maybe the `abs` shouldn't be necessary.
